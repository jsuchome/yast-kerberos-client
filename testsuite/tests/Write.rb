# encoding: utf-8

#  Write.ycp
#  Test of Kerberos::Write function
#  Author:	Jiri Suchomel <jsuchome@suse.cz>
#  $Id$
module Yast
  class WriteClient < Client
    def main
      # testedfiles: Kerberos.ycp Pam.ycp
      Yast.import "Testsuite"
      Yast.import "Kerberos"

      @READ = {
        "etc" => {
          # /etc/krb5.conf
          "krb5_conf" => {
            "s" => { "libdefaults" => {}, "SUSE.CZ" => {}, "pam" => {} }
          },
          # /etc/security/pam_*
          "security"  => {
            "section" => {
              "/etc/security/pam_unix2.conf"   => {},
              "/etc/security/pam_pwcheck.conf" => {}
            },
            "v"       => {
              "/etc/security/pam_unix2.conf" => {
                "password" => "nullok use_ldap use_krb5",
                "auth"     => "nullok use_krb5",
                "account"  => "use_krb5"
              }
            }
          }
        }
      }

      @WRITE = {}
      @EX = { "target" => { "bash_output" => { "stdout" => "" } } }

      Testsuite.Dump(
        "==== writing without any changes ================================="
      )

      Testsuite.Test(lambda { Kerberos.Write }, [{}, @WRITE, @EX], 0)

      Testsuite.Dump("==== pam enabled with krb5-ignore_unknown_principals=")

      Kerberos.pam_modified = true
      Kerberos.use_pam_krb = true
      Kerberos.ignore_unknown = true

      Testsuite.Test(lambda { Kerberos.Write }, [@READ, @WRITE, @EX], 0)

      Kerberos.pam_modified = false
      Kerberos.ignore_unknown = false

      Testsuite.Dump(
        "==== ssh modified (enabled) ======================================"
      )

      Kerberos.ssh_modified = true
      Kerberos.ssh_support = true

      Testsuite.Test(lambda { Kerberos.Write }, [{}, @WRITE, @EX], 0)

      Testsuite.Dump(
        "==== kerberos disabled, ssh support disabled, krb5.conf untouched="
      )

      Kerberos.pam_modified = true
      Kerberos.use_pam_krb = false
      Kerberos.ssh_support = false

      Testsuite.Test(lambda { Kerberos.Write }, [@READ, @WRITE, @EX], 0)

      Testsuite.Dump(
        "==== only krb5.conf modified ====================================="
      )
      Testsuite.Dump(
        "==== (all sections exist) ========================================"
      )

      Kerberos.pam_modified = false
      Kerberos.ssh_modified = false
      Kerberos.modified = true

      Kerberos.default_realm = "SUSE.CZ"
      Kerberos.kdc = "chimera.suse.cz"

      Testsuite.Test(lambda { Kerberos.Write }, [@READ, @WRITE, @EX], 0)

      Testsuite.Dump(
        "==== (pam section doesn't exist) ================================="
      )
      @READ = {
        "etc" => {
          "krb5_conf" => { "s" => { "libdefaults" => {}, "SUSE.CZ" => {} } }
        }
      }
      Kerberos.admin_server = "user_defined" # -> won't be written

      Testsuite.Test(lambda { Kerberos.Write }, [@READ, {}, @EX], 0)

      nil
    end
  end
end

Yast::WriteClient.new.main
