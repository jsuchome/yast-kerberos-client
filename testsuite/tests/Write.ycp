/**
 *  Write.ycp
 *  Test of Kerberos::Write function
 *  Jiri Suchomel <jsuchome@suse.cz>
 *  $Id$
 */

{

    // testedfiles: Kerberos.ycp PamSettings.ycp
    import "Testsuite";
    import "Kerberos";

    map READ = $[
	"etc": $[
	    // /etc/krb5.conf
	    "krb5_conf" : $[
		"s" : $[
		    "libdefaults": $[],
		    "SUSE.CZ": $[],
		    "pam" : $[],
		],
	    ],
	    // /etc/security/pam_*
	    "security": $[
		"section": $[
		    "/etc/security/pam_unix2.conf": $[],
		    "/etc/security/pam_pwcheck.conf": $[]
		],
		"v": $[
		    "/etc/security/pam_unix2.conf": $[
			"password" : "nullok use_ldap use_krb5",
			"auth"	: "nullok use_krb5",
			"account": "use_krb5"
		    ],
		]
	    ],
	]
    ];

    map WRITE = $[];

    Testsuite::Dump ("==== writing without any changes =================================");

    Testsuite::Test(``(Kerberos::Write ()), [ $[], WRITE, $[] ], 0);

    Testsuite::Dump ("==== ssh modified (enabled) ======================================");

    Kerberos::ssh_modified = true;
    Kerberos::ssh_support = true;

    Testsuite::Test(``(Kerberos::Write ()), [ $[], WRITE, $[] ], 0);

    Testsuite::Dump ("==== kerberos disabled, ssh support disabled, krb5.conf untouched=");

    Kerberos::pam_modified = true;
    Kerberos::use_pam_krb = false;
    Kerberos::ssh_support = false;

    Testsuite::Test(``(Kerberos::Write ()), [ READ, WRITE, $[] ], 0);

    Testsuite::Dump ("\nThe map with original values is always read again in");
    Testsuite::Dump ("PamSettings::RemoveValue() function, so the output is strange...\n");

    Testsuite::Dump ("==== only krb5.conf modified =====================================");
    Testsuite::Dump ("==== (all sections exist) ========================================");

    Kerberos::pam_modified = false;
    Kerberos::ssh_modified = false;
    Kerberos::modified = true;

    Kerberos::default_realm = "SUSE.CZ";
    Kerberos::kdc = "chimera.suse.cz";

    Testsuite::Test(``(Kerberos::Write ()), [ READ, WRITE, $[] ], 0);

    Testsuite::Dump ("==== (pam section doesn't exist) =================================");
    READ = $[
	"etc": $[
	    "krb5_conf" : $[
		"s" : $[
		    "libdefaults": $[],
		    "SUSE.CZ": $[],
		],
	    ]
	]
    ];
    Kerberos::kpasswd_server = "user_defined"; // -> won't be written

    Testsuite::Test(``(Kerberos::Write ()), [ READ, $[], $[] ], 0);
}
