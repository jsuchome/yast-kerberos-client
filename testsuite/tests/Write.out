Dump	==== writing without any changes =================================
Return	true
Dump	==== ssh modified (enabled) ======================================
Write	.etc.ssh.ssh_config.v."*"."GSSAPIAuthentication" "yes" true
Write	.etc.ssh.ssh_config.v."*"."GSSAPIDelegateCredentials" "yes" true
Write	.etc.ssh.ssh_config nil true
Return	true
Dump	==== kerberos disabled, ssh support disabled, krb5.conf untouched=
Execute	.target.bash_output "pam-config -d --krb5" $["stdout":""]
Write	.etc.ssh.ssh_config.v."*"."GSSAPIAuthentication" "no" true
Write	.etc.ssh.ssh_config.v."*"."GSSAPIDelegateCredentials" "no" true
Write	.etc.ssh.ssh_config nil true
Return	true
Dump	==== only krb5.conf modified =====================================
Dump	==== (all sections exist) ========================================
Write	.etc.krb5_conf.v.libdefaults.default_realm "SUSE.CZ" true
Write	.etc.krb5_conf.v.libdefaults.clockskew "300" true
Dir	.etc.krb5_conf.s: ["SUSE.CZ", "libdefaults", "pam"]
Write	.etc.krb5_conf.v."SUSE.CZ"."kdc" "chimera.suse.cz" true
Write	.etc.krb5_conf.v."SUSE.CZ"."admin_server" "chimera.suse.cz" true
Dir	.etc.krb5_conf.s: ["SUSE.CZ", "libdefaults", "pam"]
Write	.etc.krb5_conf.v.pam."ticket_lifetime" "1d" true
Write	.etc.krb5_conf.v.pam."renew_lifetime" "1d" true
Write	.etc.krb5_conf.v.pam."forwardable" "true" true
Write	.etc.krb5_conf.v.pam."proxiable" "false" true
Write	.etc.krb5_conf.v.pam."retain_after_close" "false" true
Write	.etc.krb5_conf.v.pam."minimum_uid" "1" true
Write	.etc.krb5_conf.v.pam."use_shmem" "sshd" true
Write	.etc.krb5_conf nil true
Return	true
Dump	==== (pam section doesn't exist) =================================
Write	.etc.krb5_conf.v.libdefaults.default_realm "SUSE.CZ" true
Write	.etc.krb5_conf.v.libdefaults.clockskew "300" true
Dir	.etc.krb5_conf.s: ["SUSE.CZ", "libdefaults"]
Write	.etc.krb5_conf.v."SUSE.CZ"."kdc" "chimera.suse.cz" true
Dir	.etc.krb5_conf.s: ["SUSE.CZ", "libdefaults"]
Write	.etc.krb5_conf.st.appdefaults.pam 1 true
Write	.etc.krb5_conf.v.appdefaults.pam."ticket_lifetime" "1d" true
Write	.etc.krb5_conf.v.appdefaults.pam."renew_lifetime" "1d" true
Write	.etc.krb5_conf.v.appdefaults.pam."forwardable" "true" true
Write	.etc.krb5_conf.v.appdefaults.pam."proxiable" "false" true
Write	.etc.krb5_conf.v.appdefaults.pam."retain_after_close" "false" true
Write	.etc.krb5_conf.v.appdefaults.pam."minimum_uid" "1" true
Write	.etc.krb5_conf.v.appdefaults.pam."use_shmem" "sshd" true
Write	.etc.krb5_conf nil true
Return	true
